kernel/src/util/iovec.rs

```diff
@@ -126,3 +136,195 @@ pub fn copy_iovs_from_user(start_addr: Vaddr, count: usize) -> Result<Box<[IoVec
+) -> Result<Box<[T]>> {
+    let vm_space = ctx.process.root_vmar().vm_space();
+
+    let mut v = Vec::with_capacity(count);
+    for idx in 0..count {
+        let iov = {
+            let addr = start_addr + idx * core::mem::size_of::<UserIoVec>();
+            let uiov: UserIoVec = vm_space
+                .reader(addr, core::mem::size_of::<UserIoVec>())?
+                .read_val()?;
+            IoVec::try_from(uiov)?
+        };
+
+        if iov.is_empty() {
+            continue;
+        }
+
+        let converted = convert_iovec(&iov, vm_space)?;
+        v.push(converted)
+    }
+
+    Ok(v.into_boxed_slice())
+}
+
+/// A collection of [`VmReader`]s.
+///
+/// Such readers are built from user-provided buffer, so it's always fallible.
+pub struct IoVecReader<'a>(Box<[VmReader<'a>]>);
+
+/// A collection of [`VmWriter`]s.
+///
+/// Such writers are built from user-provided buffer, so it's always fallible.
+pub struct IoVecWriter<'a>(Box<[VmWriter<'a>]>);
+
+impl<'a> IoVecReader<'a> {
+    /// Creates a new `IoVecReader` from user-provided io vec buffer.
+    pub fn copy_from_user(ctx: &'a Context<'a>, start_addr: Vaddr, count: usize) -> Result<Self> {
+        let readers = copy_iovs_and_convert(ctx, start_addr, count, IoVec::reader)?;
+        Ok(Self(readers))
+    }
+}
+
+impl<'a> IoVecWriter<'a> {
+    /// Creates a new `IoVecWriter` from user-provided io vec buffer.
+    pub fn copy_from_user(ctx: &'a Context<'a>, start_addr: Vaddr, count: usize) -> Result<Self> {
+        let writers = copy_iovs_and_convert(ctx, start_addr, count, IoVec::writer)?;
+        Ok(Self(writers))
+    }
+
+    /// Returns all writers
+    pub fn writers(&self) -> &[VmWriter] {
+        &self.0
+    }
+
+    /// Returns all mutable writers
+    pub fn writers_mut(&'a mut self) -> &mut [VmWriter] {
+        &mut self.0
+    }
+}
+
+/// Trait defining the read behavior for a collection of [`VmReader`]s.
+pub trait IoVecRead {
+    /// Reads data to the provided writer from all collected [`VmReader`]s,
+    /// accumulating total bytes read.
+    ///
+    /// This method returns [`Errno::EFAULT`]
+    /// if reading from any [`VmReader`] encounters page fault,
+    /// and no bytes is successfully read from the [`VmReader`].
+    fn read_to(&mut self, writer: &mut VmWriter<'_, Infallible>) -> Result<usize>;
+
+    /// Returns the length of data remaining to read.
+    fn len(&self) -> usize;
+
+    /// Checks if the data remaining to read is empty.
+    fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+}
+
+/// Trait defining the write behavior for a collection of [`VmWriter`]s.
+pub trait IoVecWrite {
+    /// Writes data from the provided reader to all collected [`VmWriter`]s,
+    /// accumulating total bytes written.
+    ///     
+    /// This method returns [`Errno::EFAULT`]
+    /// if reading from any [`VmReader`] encounters page fault,
+    /// and no bytes is successfully read from the [`VmReader`].
```

Comment:
```suggestion
    /// if writing to any [`VmWriter`] encounters page fault,
    /// and no bytes is successfully written to the [`VmWriter`].
```

But actually, with https://github.com/asterinas/asterinas/pull/1294#discussion_r1751921866 and https://github.com/asterinas/asterinas/pull/1294#discussion_r1751932252, I think we want something like:
```rust
pub trait IoVecRead {
    /// [..]
    ///
    /// This method returns [`Errno::EFAULT`]
    /// if a page fault occurs and no bytes are successfully read.
    fn read_to(&mut self, writer: &mut VmWriter<'_, Infallible>) -> Result<usize>;

    /// [..]
    ///
    /// This method returns [`Errno::EFAULT`]
    /// if a page fault occurs (and not all bytes are successfully read).
    fn read_exact(&mut self, writer: &mut VmWriter<'_, Infallible>) -> Result<usize>;
}

pub trait IoVecWrite {
    /// [..]
    ///
    /// This method returns [`Errno::EFAULT`]
    /// if a page fault occurs and no bytes are successfully written.
    fn write_from(&mut self, reader: &mut VmReader<'_, Infallible>) -> Result<usize>;

    /// [..]
    ///
    /// This method returns [`Errno::EFAULT`]
    /// if a page fault occurs (and not all bytes are successfully written).
    fn write_all(&mut self, reader: &mut VmReader<'_, Infallible>) -> Result<usize>;
}
```

Then we'll have the following users:
 - `read_to/write_from`: pipe `read`/`write`
 - `read_exact/write_all`: TCP/UDP `send`/`recv`

Note that for UNIX sockets they should go through `read_exact`/`write_all`, but currently have to go through `read_to`/`write_from` because their implementation is based on pipes.
 - Leaving this unfixed (perhaps adding a FIXME somewhere if there's a good place to note it) is fine, since it requires more code changes, and it's doubtful that the code changes are worth it.

Reply:
The trait has been refactored as suggested. 

I have add two FIXME on `Connected::try_read` and `Connected::try_write` about the unix socket. 

> write_from: pipe write, TCP recv

It seems `TCP recv` should also call `write_all`. A partially bad receive buffer will also returns EFAULT.

```C
char recv_buffer[4096] = { 0 };
iov[0].iov_base = recv_buffer;
iov[0].iov_len = 1;
// Bad buffer
iov[1].iov_base = (char *)1;
iov[1].iov_len = 1;
msg.msg_iovlen = 2;
TEST_ERRNO(recvmsg(sk_bound, &msg, 0), EFAULT);
```

Reply:
> I have add two FIXME on Connected::try_read and Connected::try_write about the unix socket.

Sorry, this does not seem necessary according to the comment above.

Actually, it just means that `read_to`/`write_from` doesn't seem to have any users (I originally thought it might be useful for pipes, but now it turns out they just use `read_exact` and `write_all` instead). This is sad. We can just remove it, or leave it for future users (in case something behaves like the files in #711).

Reply:
I have removed `read_to/write_from`, since they don't have any use yet.